From: Mike Snitzer <snitzer@redhat.com>
Date: Mon, 11 Jan 2010 21:30:53 -0500
Subject: [block] bdev_stack_limits wrapper
Message-id: <1263245455-7547-4-git-send-email-snitzer@redhat.com>
Patchwork-id: 22412
O-Subject: [RHEL6 PATCH 3/5] block: bdev_stack_limits wrapper
Bugzilla: 554718
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ 549766

backport from linux-2.6-block:

commit 17be8c245054b9c7786545af3ba3ca4e54cd4ad9
Author: Martin K. Petersen <martin.petersen@oracle.com>
Date:   Mon Jan 11 03:21:49 2010 -0500

    block: bdev_stack_limits wrapper

    DM does not want to know about partition offsets.  Add a partition-aware
    wrapper that DM can use when stacking block devices.

Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Reviewed-by: Alasdair G Kergon <agk@redhat.com>

diff --git a/block/blk-settings.c b/block/blk-settings.c
index 7493182..9368756 100644
--- a/block/blk-settings.c
+++ b/block/blk-settings.c
@@ -639,6 +639,28 @@ int blk_stack_limits(struct queue_limits *t, struct queue_limits *b,
 EXPORT_SYMBOL(blk_stack_limits);
 
 /**
+ * bdev_stack_limits - adjust queue limits for stacked drivers
+ * @t:	the stacking driver limits (top device)
+ * @bdev:  the component block_device (bottom)
+ * @start:  first data sector within component device
+ *
+ * Description:
+ *    Merges queue limits for a top device and a block_device.  Returns
+ *    0 if alignment didn't change.  Returns -1 if adding the bottom
+ *    device caused misalignment.
+ */
+int bdev_stack_limits(struct queue_limits *t, struct block_device *bdev,
+		      sector_t start)
+{
+	struct request_queue *bq = bdev_get_queue(bdev);
+
+	start += get_start_sect(bdev);
+
+	return blk_stack_limits(t, &bq->limits, start << 9);
+}
+EXPORT_SYMBOL(bdev_stack_limits);
+
+/**
  * disk_stack_limits - adjust queue limits for stacked drivers
  * @disk:  MD/DM gendisk (top)
  * @bdev:  the underlying block device (bottom)
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index 001c9e6..a6703c0 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -946,6 +946,8 @@ extern void blk_queue_io_opt(struct request_queue *q, unsigned int opt);
 extern void blk_set_default_limits(struct queue_limits *lim);
 extern int blk_stack_limits(struct queue_limits *t, struct queue_limits *b,
 			    sector_t offset);
+extern int bdev_stack_limits(struct queue_limits *t, struct block_device *bdev,
+			    sector_t offset);
 extern void disk_stack_limits(struct gendisk *disk, struct block_device *bdev,
 			      sector_t offset);
 extern void blk_queue_stack_limits(struct request_queue *t, struct request_queue *b);
